bin/gollum-migrate-tags

```diff
@@ -0,0 +1,242 @@
+        next '' if m_end.length < m_start.length
+        lang = m_lang ? m_lang.strip.split.first : nil
+        cache_codeblock($~.to_s)
+      end  
+    end
+
+    data.gsub!(/^([ ]{0,3})``` ?([^\r\n]+)?\r?\n(.+?)\r?\n[ ]{0,3}```[ \t]*\r?$/m) do
+      cache_codeblock($~.to_s)
+    end
+    data
+  end
+  
+  def process(data)
+    return data if data.nil? || data.size.zero? || @map.size.zero?
+    @map.each do |id, block| ## Just put the code blocks back in verbatim
+      data.gsub!(id, block)
+    end
+    data
+  end
+  
+  def cache_codeblock(block)
+    id = "#{open_pattern}#{Digest::SHA1.hexdigest(block)}#{close_pattern}"
+    @map[id] = block
+    id
+  end
+end
+  
+class ::Gollum::Filter::TagMigrator < Gollum::Filter::Tags
+  def process_tag(tag)
+    link_part, extra = parse_tag_parts(tag)
+    orig_tag = %{[[#{tag}]]}
+    return orig_tag if link_part.nil?
+    
+    img_args = extra ? [extra, link_part] : [link_part]
+    mime = MIME::Types.type_for(::File.extname(img_args.first.to_s)).first
+    
+    # For any kind of tag other than an internal link: just return the tag.
+    if tag =~ /^_TOC_/ || link_part =~ /^_$/ || link_part =~ /^#{INCLUDE_TAG}/ || (mime && mime.content_type =~ /^image/) || process_external_link_tag(link_part, extra) || process_file_link_tag(link_part, extra)
+      return orig_tag
+    end
+
+    # It's an internal Page link tag. Try to find the file/page it links to
+    link = link_part
+    page = find_page_from_path(link)
+    anchor = nil
+    
+    if page.nil? # No match yet, now try finding the page with anchor removed
+      if pos = link.rindex('#')
+        anchor = link[pos..-1]
+        link  = link[0...pos]
+      end
+
+      if link.empty? && anchor # Internal anchor link, don't search for the page but return the original tag
+        return orig_tag
+      end
+
+      page  = find_page_from_path(link)
+    end
+
+    if page
+      # Great, the link is not broken. Return the original tag.
+      return orig_tag
+    else
+      possibles = find_linked(link)
+      if possibles.empty?
+        log(:info, "Found no candidates for broken link: #{orig_tag}")
+        return orig_tag
+      else
+        if possibles.size > 1
+          log(:empty)
+          log(:warn, "Found multiple possibilities for the link '#{orig_tag}':")
+          possibles.map! {|p| Pathname.new(p)}
+          possibles.sort!
+          possibles.each{|p| log(:none, "* #{p}")}
+          log(:warn,"Picking #{possibles.first}")
+          log(:empty)
+        end
+        pick = possibles.first
+        return tag_for_pick(pick, orig_tag, extra, anchor, @markup.page.path)
+      end
+        
+    end
+  end
+  
+  private
+  
+  def tag_for_pick(pick, orig_tag, extra, anchor, linking_page_path)
+    pick = if defined?(PREFER_RELATIVE)
+      overlapping_path = Pathname.new(linking_page_path).dirname.to_s
+      relative_path = pick.to_s.match(/^\/#{overlapping_path}\/(.+)/)
```

Comment:
Given two files:

* `Subdir/Foo`
* `Subdir/Subsub/Bar`

It checks whether `Bar` is under the same directory as `Foo`, i.e. if they have an 'overlapping path'. In this case, the answer is yes: `Subdir/`. The capture group then contains a relative path to `Bar` from the overlapping path: `Subsub/Bar`.

I don't see a way of doing this with Pathname, unfortunately. It would be great to have such a method for determining whethere one path is 'under' another.

Reply:
How is this?
```ruby
Pathname.new('Subdir/Subsub/Bar').relative_path_from(Pathname.new('Subdir/Foo').dirname)
 => #<Pathname:Subsub/Bar> 
```

Reply:
😍